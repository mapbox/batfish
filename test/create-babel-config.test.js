'use strict';

const path = require('path');
const createBabelConfig = require('../src/node/create-babel-config');

const abs = (x) => path.join(__dirname, x);

const defaultBatfishConfig = () => ({
  babelPlugins: [],
  babelPresets: [],
  jsxtremeMarkdownOptions: {},
  browserslist: 'mock-browserslist',
  production: true
});

const relativizeBabelSetting = (p) => {
  if (!Array.isArray(p)) {
    return `./${path.relative(__dirname, p)}`;
  }
  return [`./${path.relative(__dirname, p[0])}`, p[1]];
};

describe('createBabelConfig', () => {
  test('with defaults', () => {
    const actual = createBabelConfig(defaultBatfishConfig());
    expect(actual.presets.map(relativizeBabelSetting)).toMatchSnapshot();
    expect(actual.plugins.map(relativizeBabelSetting)).toMatchSnapshot();
  });

  test('with Node target', () => {
    const actual = createBabelConfig(defaultBatfishConfig(), {
      target: 'node'
    });
    expect(actual.presets.map(relativizeBabelSetting)).toMatchSnapshot();
  });

  test('with Batfish config that affects Babel', () => {
    const actual = createBabelConfig(
      Object.assign(defaultBatfishConfig(), {
        babelPlugins: [abs('foo/bar'), abs('baz')],
        babelPresets: [abs('a/b'), abs('c')],
        jsxtremeMarkdownOptions: {
          remarkPlugins: ['remark-plugin-a'],
          rehypePlugins: ['rehype-plugin-a', 'rehype-plugin-b']
        }
      })
    );
    expect(actual.presets.map(relativizeBabelSetting)).toMatchSnapshot();
    expect(actual.plugins.map(relativizeBabelSetting)).toMatchSnapshot();
  });

  test('non-false devBrowserslist is used in non-production', () => {
    const actual = createBabelConfig(
      Object.assign(defaultBatfishConfig(), {
        production: false,
        devBrowserslist: 'mock-dev-browserslist'
      })
    );
    expect(actual.presets.map(relativizeBabelSetting)).toMatchSnapshot();
  });

  test('false devBrowserslist is not used in non-production', () => {
    const actual = createBabelConfig(
      Object.assign(defaultBatfishConfig(), {
        production: false,
        devBrowserslist: false
      })
    );
    expect(actual.presets.map(relativizeBabelSetting)).toMatchSnapshot();
  });

  test('non-false devBrowserslist is not used in production', () => {
    const actual = createBabelConfig(
      Object.assign(defaultBatfishConfig(), {
        production: true,
        devBrowserslist: 'mock-dev-browserslist'
      })
    );
    expect(actual.presets.map(relativizeBabelSetting)).toMatchSnapshot();
  });
});
