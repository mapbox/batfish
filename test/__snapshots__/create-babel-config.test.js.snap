// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBabelConfig false devBrowserslist is not used in non-production 1`] = `
Array [
  Array [
    "./../node_modules/babel-preset-env/lib/index.js",
    Object {
      "modules": false,
      "targets": Object {
        "browsers": "mock-browserslist",
      },
      "useBuiltIns": true,
    },
  ],
  "./../node_modules/babel-preset-react/lib/index.js",
]
`;

exports[`createBabelConfig non-false devBrowserslist is not used in production 1`] = `
Array [
  Array [
    "./../node_modules/babel-preset-env/lib/index.js",
    Object {
      "modules": false,
      "targets": Object {
        "browsers": "mock-browserslist",
      },
      "useBuiltIns": true,
    },
  ],
  "./../node_modules/babel-preset-react/lib/index.js",
]
`;

exports[`createBabelConfig non-false devBrowserslist is used in non-production 1`] = `
Array [
  Array [
    "./../node_modules/babel-preset-env/lib/index.js",
    Object {
      "modules": false,
      "targets": Object {
        "browsers": "mock-dev-browserslist",
      },
      "useBuiltIns": true,
    },
  ],
  "./../node_modules/babel-preset-react/lib/index.js",
]
`;

exports[`createBabelConfig with Batfish config that affects Babel 1`] = `
Array [
  Array [
    "./../node_modules/babel-preset-env/lib/index.js",
    Object {
      "modules": false,
      "targets": Object {
        "browsers": "mock-browserslist",
      },
      "useBuiltIns": true,
    },
  ],
  "./../node_modules/babel-preset-react/lib/index.js",
  "./a/b",
  "./c",
]
`;

exports[`createBabelConfig with Batfish config that affects Babel 2`] = `
Array [
  "./../node_modules/babel-plugin-syntax-dynamic-import/lib/index.js",
  "./../node_modules/babel-plugin-transform-class-properties/lib/index.js",
  "./../node_modules/babel-plugin-transform-object-rest-spread/lib/index.js",
  Array [
    "./../node_modules/@mapbox/babel-plugin-transform-jsxtreme-markdown/index.js",
    Object {
      "packageName": "@mapbox/batfish/modules/md",
      "rehypePlugins": Array [
        "rehype-plugin-a",
        "rehype-plugin-b",
      ],
      "remarkPlugins": Array [
        "remark-plugin-a",
      ],
    },
  ],
  "./foo/bar",
  "./baz",
  "./../babel-plugin-transform-react-remove-prop-types",
]
`;

exports[`createBabelConfig with Node target 1`] = `
Array [
  Array [
    "./../node_modules/babel-preset-env/lib/index.js",
    Object {
      "modules": false,
      "targets": Object {
        "node": "current",
      },
      "useBuiltIns": true,
    },
  ],
  "./../node_modules/babel-preset-react/lib/index.js",
]
`;

exports[`createBabelConfig with defaults 1`] = `
Array [
  Array [
    "./../node_modules/babel-preset-env/lib/index.js",
    Object {
      "modules": false,
      "targets": Object {
        "browsers": "mock-browserslist",
      },
      "useBuiltIns": true,
    },
  ],
  "./../node_modules/babel-preset-react/lib/index.js",
]
`;

exports[`createBabelConfig with defaults 2`] = `
Array [
  "./../node_modules/babel-plugin-syntax-dynamic-import/lib/index.js",
  "./../node_modules/babel-plugin-transform-class-properties/lib/index.js",
  "./../node_modules/babel-plugin-transform-object-rest-spread/lib/index.js",
  Array [
    "./../node_modules/@mapbox/babel-plugin-transform-jsxtreme-markdown/index.js",
    Object {
      "packageName": "@mapbox/batfish/modules/md",
      "rehypePlugins": undefined,
      "remarkPlugins": undefined,
    },
  ],
  "./../babel-plugin-transform-react-remove-prop-types",
]
`;
